### PR TITLE
Add: Masked Autoencoders Are Scalable Vision Learners

### PR DESCRIPTION
**Pull request by ariG23498**
Going ahead with the notion of "reconstruction as a pretext task" He et. al. in their [paper](https://arxiv.org/abs/2111.06377) have ablated on masking patches and predicting the patches with ViT autoencoders as a pretraining task. We (with @sayakpaul) have implemented the paper and have given thought to even the most minute details.

With just 100 epochs of pre-training and a fairly lightweight Autoencoder architecture we achieve **44.25%** accuracy with linear probing on the CIFAR-10 dataset.

[Colab notebook](https://colab.research.google.com/drive/1DYBcGuCfDV7d6GcsOx3BxsIVSuUtu4fT?usp=sharing)

### CODE REVIEW

**Comment by innat** [comment]
@ariG23498 
In your colab, I think you need to add the following in the callback class.

```python
def on_epoch_end(self, epoch, logs=None):
   if self.epoch_interval and epoch % self.epoch_interval == 0:
        test_augmeneted_images = self.model.test_augmentation_model(test_images) < ---- here
        test_patches = self.model.patch_layer(test_augmeneted_images)
        ...
```

**Comment by ariG23498** [comment]
@innat thanks for the catch!

File: examples/vision/masked_image_modeling.py [hunk 1]
```diff
@@ -0,0 +1,935 @@
...
+        self.test_augmentation_model = test_augmentation_model
+        self.patch_layer = patch_layer
+        self.patch_encoder = patch_encoder
+        self.encoder = encoder
+        self.decoder = decoder
+
+    def calculate_loss(self, images, test=False):
+        # Augment the input images.
+        if test:
+            augmeneted_images = self.test_augmentation_model(images)
```
Comment (fchollet) [inline examples/vision/masked_image_modeling.py:569]: "augmented" (typo is repeated in a number of places)

File: examples/vision/masked_image_modeling.py [hunk 2]
```diff
@@ -0,0 +1,935 @@
...
+        # Update the trackers.
+        self.compiled_metrics.update_state(loss_patch, loss_output)
+        return {m.name: m.result() for m in self.metrics}
+
+
+"""
+## Model initialization
+"""
+
+keras.backend.clear_session()
```
Comment (fchollet) [inline examples/vision/masked_image_modeling.py:639]: Why is this call necessary?

File: examples/vision/masked_image_modeling.py [hunk 3]
```diff
@@ -0,0 +1,935 @@
...
+* [SimCLR](https://keras.io/examples/vision/semisupervised_simclr/)
+* [NNCLR](https://keras.io/examples/vision/nnclr)
+* [SimSiam](https://keras.io/examples/vision/simsiam)
+
+This idea of using BERT flavored pretraining in computer vision was also explored in
+[Selfie](https://arxiv.org/abs/1906.02940), but it could not demonstrate strong results.
+Another concurrent work that explores the idea of masked image modeling is
+[SimMIM](https://arxiv.org/abs/2111.09886). Finally, as a fun fact, we, the authors of
+this example also explored the idea of ["reconstruction as a pretext
+task"](https://i.ibb.co/k5CpwDX/image.png) in 2020 but we could not prevent the network
```
Comment (fchollet) [inline examples/vision/masked_image_modeling.py:927]: Please make sure markdown links fit on a single line

**Review by fchollet — COMMENTED** [review]
Great example, I love it. Very good code quality and very readable. Thanks for the PR!

**Comment by ariG23498** [comment]
@fchollet thank you for the review. We have made the necessary changes.

**Comment by ariG23498** [comment]
EDIT: Updating the masking ratio from 60% to 75% gave us an increase in the linear probing accuracy on test set from **44.25%** to **46.70%**. This finding aligns with the literature.

CC: @sayakpaul

**Comment by beresandras** [comment]
Very nice example! One note, can you consider adding the following two works from earlier this year to the related works section (I am not affiliated with any of them)?
- [SiT](https://arxiv.org/abs/2104.03602): trains ViT with a (non-patchwise) denoising objective (MAE not citing them has already caused some [controversy](https://twitter.com/ducha_aiki/status/1460014465253052418))
- [BEiT](https://arxiv.org/abs/2106.08254): trains ViT with the same masked autoencoding objective, but using discrete VAE tokens instead of in pixel-space.

**Comment by sayakpaul** [comment]
Thank you for your comments, @beresandras. For a more exhaustive literature survey interested readers could check out the MAE paper itself and the other two papers we have mentioned.

**Comment by innat** [comment]
@ariG23498  @sayakpaul  
Is it possible to add TPU training settings as well (additionally)?

**Comment by sayakpaul** [comment]
For now, we will keep it as it is.

**Comment by innat** [comment]
@sayakpaul definitely. 

[ However, just from curiosity, I was trying to run the above colab on TPU but ended up with some issue, [colab link](https://drive.google.com/file/d/1w5aQGLBXwYlao-p1BAlixqgk5cW8QSY_/view?usp=sharing). I also tried with `tf.config.set_soft_device_placement(True)` but it gave another issue. Actually, these are quite similar to an old issue that we faced previously, [here](https://github.com/keras-team/keras/issues/15577). (Though I found some workaround but not general.) ]

**Comment by sayakpaul** [comment]
Yes, those issues are the same ones we faced during the development process. Thanks for letting us know.

**Comment by sayakpaul** [comment]
@fchollet a gentle reminder on this PR. 

We were checking the file generation process for this PR and it errored out complaining that the code has 482 lines. We thought it might be a good time to discuss what do we do about that.

**Review by fchollet — APPROVED** [review]
Thanks for the update -- everything is looking good! Please add the generated files.

**Comment by fchollet** [comment]
> We were checking the file generation process for this PR and it errored out complaining that the code has 482 lines. We thought it might be a good time to discuss what do we do about that.

That's ok, you can temporarily disable the check (by increasing the limit).

**Comment by sayakpaul** [comment]
@fchollet thank you! Just added all the files.

**Comment by fchollet** [comment]
Thank you for the contribution!
